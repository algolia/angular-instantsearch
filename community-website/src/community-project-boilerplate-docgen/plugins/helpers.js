const highlight = require('./syntaxHighlighting.js');
const md = require('./mdRenderer.js');

// this plugin provides ATM one helper to easily compute the publicPath of assets
module.exports = function helpers(filenames, metalsmith, cb) {
  // eslint-disable-next-line no-param-reassign
  metalsmith.metadata().h = {
    markdown(src) {
      return md.render(src);
    },
    highlight(src, opts) {
      const lang = opts && opts.lang;
      return highlight(src, lang);
    },
    maybeActive(navPath, singlePathOrArrayOfPaths) {
      const pathsToTest = [].concat(singlePathOrArrayOfPaths);
      return pathsToTest.some(pathToTest => navPath.indexOf(pathToTest) === 0)
        ? 'active'
        : '';
    },
  };

  cb();
};
