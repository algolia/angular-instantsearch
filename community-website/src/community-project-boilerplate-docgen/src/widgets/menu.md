---
title: Widgets - Menu
layout: widget.pug
canonical: https://www.algolia.com/doc/api-reference/widgets/menu/angular/
---

# Menu

## Description

The Menu component displays a menu that lets the user choose a single value for a specific attribute.

## Options

| Attribute            | Type                  | Description
| -                    | -                     | -
| `attribute`          | `string`              | Name of the attribute for faceting (eg. "free_shipping")
| `limit?`             | `number`              | How many facets values to retrieve
| `showMoreLimit?`     | `number`              | Bigger than `limit` if the component should display a button that will expand the number of items
| `showMoreLabel?`     | `string`              | Label of the show more button
| `showLessLabel?`     | `string`              | Label of the show less button
| `sortBy?`            | `string[] / Function` | How to sort facet values
| `autoHideContainer?` | `boolean`             | Hides the menu if there's no item to display
| `transformItems?`    | `(items) => items`    | Function to modify the items being displayed, e.g. for filtering or sorting them

## Code example

```html
<ais-menu
  attribute="categories"
  [limit]="3"
  [showMoreLimit]="10"
  [transformItems]="translate"
>
</ais-menu>
```
```ts
export class AppComponent {
  // ...
  translate(items) {
    return items.map(item => ({
      ...item,
      highlighted: myTranslateFunc(item.highlighted)
    }));
  }
}
```
