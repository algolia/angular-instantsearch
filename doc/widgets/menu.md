## `<ng-ais-menu />`

The Menu component displays a menu that lets the user choose a single value for a specific attribute.


| Attributes           | Type                  | Description
| -                    | -                     | -
| `attributeName`      | `string`              | Name of the attribute for faceting (eg. "free_shipping")
| `limitMin?`          | `number`              | How many facets values to retrieve
| `limitMax?`          | `number`              | Bigger than `limit` if the component should display a button that will expand the number of items
| `showMoreLabel?`     | `string`              | Label of the show more button
| `showLessLabel?`     | `string`              | Label of the show less button
| `sortBy?`            | `string[] / Function` | How to sort facet values
| `autoHideContainer?` | `boolean`             | Hides the menu if there's no item to display
