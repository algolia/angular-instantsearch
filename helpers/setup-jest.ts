import "jest-preset-angular";

// stupid workaround for TS
import * as helper from "algoliasearch-helper";
require("algoliasearch-helper").default = helper;

const mock = () => {
  let storage = {};
  return {
    getItem(key) {
      return storage[key] || null;
    },
    setItem(key, value) {
      storage[key] = value;
    },
    removeItem(key) {
      delete storage[key];
    },
    clear() {
      storage = {};
    }
  };
};

Object.defineProperty(window, "localStorage", { value: mock() });
Object.defineProperty(window, "sessionStorage", { value: mock() });
Object.defineProperty(window, "getComputedStyle", {
  value: () => ["-webkit-appearance"]
});
