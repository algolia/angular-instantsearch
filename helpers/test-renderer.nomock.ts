import { Component, ViewChild } from '@angular/core';
import { TestBed } from '@angular/core/testing';

import { NgAisInstantSearchModule } from '../src/instantsearch/instantsearch.module';

// same as test-renderer, but without mocking InstantSearch
// In the future, we want to migrate all tests to this

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations,
  methods = {},
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
  methods?: {};
}) {
  return function(state?: {}, isFirstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined,
        methods,
      },
      isFirstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state,
    methods,
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
    methods?: {};
  },
  isFirstRender = false
) {
  @Component({
    template: `
      <ais-instantsearch [config]="config">
        ${template}
      </ais-instantsearch>
    `,
  })
  class TestContainer {
    config = {
      searchClient: {
        search(requests) {
          return Promise.resolve({
            results: requests.map(() => ({ hits: [] })),
          });
        },
      },
      indexName: 'instant_search',
    };
    @ViewChild(TestedWidget) testedWidget;
    constructor() {
      Object.keys(methods).forEach(methodName => {
        this[methodName] = methods[methodName];
      });
    }
  }

  TestBed.configureCompiler({
    preserveWhitespaces: false,
  } as any).configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || []),
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])],
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, isFirstRender);
    fixture.detectChanges();
  }

  return fixture;
}
