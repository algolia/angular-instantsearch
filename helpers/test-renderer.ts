import { Component, ViewChild } from "@angular/core";
import { TestBed } from "@angular/core/testing";

import { NgAisInstantSearchModule } from "../src/instantsearch/instantsearch.module";

// mock component, dont create a real instantearch instance
jest.mock("../src/base-widget");
jest.mock("../src/instantsearch/instantsearch");

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
}) {
  return function(state?: {}, firstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined
      },
      firstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
  },
  firstRender = false
) {
  @Component({
    template: `
      <ng-ais-instantsearch>
        ${template}
      </ng-ais-instantsearch>
    `
  })
  class TestContainer {
    @ViewChild(TestedWidget) testedWidget;
  }

  TestBed.configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || [])
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])]
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, firstRender);
    fixture.detectChanges();
  }

  return fixture;
}
