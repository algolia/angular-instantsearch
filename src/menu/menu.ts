import { Component, Input, Inject, forwardRef } from '@angular/core';

import { connectMenu } from 'instantsearch.js/es/connectors';
import { BaseWidget } from '../base-widget';
import { NgAisInstantSearch } from '../instantsearch/instantsearch';
import { noop } from '../utils';

export type MenuItem = {
  value: string;
  label: string;
  count: number;
  isRefined: boolean;
};

export type MenuState = {
  items: MenuItem[];
  refine: Function;
  createURL: Function;
  isShowingMore: boolean;
  canToggleShowMore: boolean;
  toggleShowMore: Function;
};

@Component({
  selector: 'ais-menu',
  template: `
    <div
      [class]="cx()"
      *ngIf="!isHidden"
    >
      <ul [class]="cx('list')">
        <li
          [class]="getItemClass(item)"
          *ngFor="let item of state.items"
          (click)="handleClick($event, item.value)"
        >
          <a
            href="{{state.createURL(item.value)}}"
            [class]="cx('link')"
            (click)="handleClick($event, item.value)"
          >
            <span [class]="cx('label')">{{item.label}}</span>
            <span [class]="cx('count')">{{item.count}}</span>
          </a>
        </li>
      </ul>

      <button
        *ngIf="showMore"
        (click)="state.toggleShowMore()"
        [class]="showMoreClass"
        [disabled]="!state.canToggleShowMore"
      >
        {{state.isShowingMore ? showLessLabel : showMoreLabel}}
      </button>
    </div>
  `,
})
export class NgAisMenu extends BaseWidget {
  // rendering options
  @Input() public showMoreLabel: string = 'Show more';
  @Input() public showLessLabel: string = 'Show less';

  // instance options
  @Input() public attribute: string;
  @Input() public showMore?: boolean;
  @Input() public limit?: number;
  @Input() public showMoreLimit?: number;
  @Input() public sortBy?: string[] | ((item: object) => number);
  @Input()
  public transformItems?: <U extends MenuItem>(items: MenuItem[]) => U[];

  public state: MenuState = {
    items: [],
    refine: noop,
    createURL: noop,
    isShowingMore: false,
    canToggleShowMore: false,
    toggleShowMore: noop,
  };

  get isHidden() {
    return this.state.items.length === 0 && this.autoHideContainer;
  }

  get showMoreClass() {
    let className = this.cx('showMore');

    if (!this.state.canToggleShowMore) {
      className = `${className} ${this.cx('showMore', 'disabled')}`;
    }

    return className;
  }

  constructor(
    @Inject(forwardRef(() => NgAisInstantSearch))
    public instantSearchParent: any
  ) {
    super('Menu');
  }

  public ngOnInit() {
    this.createWidget(connectMenu, {
      attribute: this.attribute,
      showMore: this.showMore,
      limit: this.limit,
      showMoreLimit: this.showMoreLimit,
      sortBy: this.sortBy,
      transformItems: this.transformItems,
    });

    super.ngOnInit();
  }

  handleClick(event: MouseEvent, value: string) {
    event.preventDefault();
    event.stopPropagation();

    this.state.refine(value);
  }
}
