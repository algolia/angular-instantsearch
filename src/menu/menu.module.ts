import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { NgAisMenu } from './menu';

@NgModule({
  declarations: [NgAisMenu],
  entryComponents: [NgAisMenu],
  exports: [NgAisMenu],
  imports: [CommonModule],
})
export class NgAisMenuModule {}
