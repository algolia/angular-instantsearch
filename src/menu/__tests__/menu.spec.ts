import { createRenderer } from "../../../helpers/test-renderer";
import { NgAisMenu } from "../menu";

const defaultState = {
  canRefine: true,
  canToggleShowMore: false,
  createURL: jest.fn(),
  isShowingMore: false,
  items: [
    { value: "foo", label: "foo", count: 2, isRefined: true },
    { value: "bar", label: "bar", count: 3 },
    { value: "foobar", label: "foobar", count: 4 }
  ],
  refine: jest.fn(),
  toggleShowMore: jest.fn()
};

const render = createRenderer({
  defaultState,
  template: "<ng-ais-menu></ng-ais-menu>",
  TestedWidget: NgAisMenu
});

describe("Menu", () => {
  it("renders markup without state", () => {
    const fixture = render();
    expect(fixture).toMatchSnapshot();
  });

  it("renders markup with state", () => {
    const fixture = render({});
    expect(fixture).toMatchSnapshot();
  });

  it("should call refine() when clicking on an element", () => {
    const refine = jest.fn();
    const fixture = render({ refine });

    const [firstItem] = fixture.debugElement.nativeElement.querySelectorAll(
      "li"
    );
    firstItem.click();

    expect(refine).toHaveBeenCalled();
    expect(refine).toHaveBeenCalledWith(defaultState.items[0].value);
  });

  it("should call toggleShowMore() when possible", () => {
    const toggleShowMore = jest.fn();
    const fixture = render({ toggleShowMore, canToggleShowMore: true });

    fixture.componentInstance.testedWidget.limit = 3;
    fixture.componentInstance.testedWidget.showMoreLimit = 4;
    fixture.detectChanges();

    fixture.debugElement.nativeElement.querySelector("button").click();

    expect(toggleShowMore).toHaveBeenCalled();
  });

  it("should apply `transformItems` if specified", () => {
    const fixture = render({});
    fixture.componentInstance.testedWidget.transformItems = items =>
      items.map(item => ({ ...item, label: `transformed - ${item.label}` }));
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it("should be hidden with `autoHideContainer`", () => {
    const fixture = render({ items: [] });
    fixture.componentInstance.testedWidget.autoHideContainer = true;
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });
});
