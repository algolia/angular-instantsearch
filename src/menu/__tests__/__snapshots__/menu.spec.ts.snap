// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu renders markup with state 1`] = `
<ng-ais-menu
  cx={[Function Function]}
  limitMin={[Function Number]}
  platformId={[Function String]}
  searchInstance={[Function NgAisInstance]}
  showLessLabel={[Function String]}
  showMoreLabel={[Function String]}
  state={[Function Object]}
  updateState={[Function Function]}
>
  
      <div
    class="ais-Menu"
  >
    
        
    
    <div
      class="ais-Menu-body ais-body"
    >
      
          
      <ul
        class="ais-Menu-list"
      >
        
            
        
        <li
          class="ais-Menu-item ais-Menu-item--selected"
        >
          
              
          <a
            href=""
          >
            
                foo
                
            <span
              class="ais-Menu-count"
            >
              2
            </span>
            
              
          </a>
          
            
        </li>
        <li
          class="ais-Menu-item"
        >
          
              
          <a
            href=""
          >
            
                bar
                
            <span
              class="ais-Menu-count"
            >
              3
            </span>
            
              
          </a>
          
            
        </li>
        <li
          class="ais-Menu-item"
        >
          
              
          <a
            href=""
          >
            
                foobar
                
            <span
              class="ais-Menu-count"
            >
              4
            </span>
            
              
          </a>
          
            
        </li>
        
          
      </ul>
      
  
          
      
      
        
    </div>
    
      
  </div>
    
</ng-ais-menu>
`;

exports[`Menu renders markup without state 1`] = `
<ng-ais-menu
  cx={[Function Function]}
  limitMin={[Function Number]}
  platformId={[Function String]}
  searchInstance={[Function NgAisInstance]}
  showLessLabel={[Function String]}
  showMoreLabel={[Function String]}
  state={[Function Object]}
  updateState={[Function Function]}
>
  
      <div
    class="ais-Menu"
  >
    
        
    
    <div
      class="ais-Menu-body ais-body"
    >
      
          
      <ul
        class="ais-Menu-list"
      >
        
            
        
        
          
      </ul>
      
  
          
      
      
        
    </div>
    
      
  </div>
    
</ng-ais-menu>
`;

exports[`Menu should apply \`transformItems\` if specified 1`] = `
<ng-ais-menu
  cx={[Function Function]}
  limitMin={[Function Number]}
  platformId={[Function String]}
  searchInstance={[Function NgAisInstance]}
  showLessLabel={[Function String]}
  showMoreLabel={[Function String]}
  state={[Function Object]}
  transformItems={[Function Function]}
  updateState={[Function Function]}
>
  
      <div
    class="ais-Menu"
  >
    
        
    
    <div
      class="ais-Menu-body ais-body"
    >
      
          
      <ul
        class="ais-Menu-list"
      >
        
            
        
        <li
          class="ais-Menu-item ais-Menu-item--selected"
        >
          
              
          <a
            href=""
          >
            
                transformed - foo
                
            <span
              class="ais-Menu-count"
            >
              2
            </span>
            
              
          </a>
          
            
        </li>
        <li
          class="ais-Menu-item"
        >
          
              
          <a
            href=""
          >
            
                transformed - bar
                
            <span
              class="ais-Menu-count"
            >
              3
            </span>
            
              
          </a>
          
            
        </li>
        <li
          class="ais-Menu-item"
        >
          
              
          <a
            href=""
          >
            
                transformed - foobar
                
            <span
              class="ais-Menu-count"
            >
              4
            </span>
            
              
          </a>
          
            
        </li>
        
          
      </ul>
      
  
          
      
      
        
    </div>
    
      
  </div>
    
</ng-ais-menu>
`;

exports[`Menu should be hidden with \`autoHideContainer\` 1`] = `
<ng-ais-menu
  autoHideContainer={[Function Boolean]}
  cx={[Function Function]}
  limitMin={[Function Number]}
  platformId={[Function String]}
  searchInstance={[Function NgAisInstance]}
  showLessLabel={[Function String]}
  showMoreLabel={[Function String]}
  state={[Function Object]}
  updateState={[Function Function]}
>
  
      <div
    class="ais-Menu"
  >
    
        
    
    
      
  </div>
    
</ng-ais-menu>
`;
